### PR TITLE
Initial commit of lattice mode

### PR DESCRIPTION
**Pull request by autodidaddict**
While 99.9% of lattice mode should "just work" , there are a couple of caveats:

1) Once you compile in lattice mode, you'll need a NATS server running bound to the loopback adapter at 127.0.0.1 on the default port with no authentication (update to this PR with credential support is coming)

2) While there's no breaking changes, `bindings` are now more "real" things and the act of creating one (via `bind_actor` or loading a manifest) is more transactional than before. With or without lattice mode enabled, an actor _cannot talk to a provider_ that it has not yet bound to, even if the binding contains no configuration data.

For more information on what lattice is and how it works, check out [this PR](https://github.com/wascc/wascc-dev-site/pull/18) for the doc site.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -196,16 +194,35 @@ impl WasccHost {
...
+            None,
+            self.bus.clone(),
+            self.middlewares.clone(),
+            self.caps.clone(),
+            self.bindings.clone(),
+            self.claims.clone(),
+            self.terminators.clone(),
         )?;
         wg.wait();
+        if actor.capabilities().contains(&"wascc:extras".into()) {
```
Comment (ewbankkit) [inline src/lib.rs]: Constant for `"wascc:extras"`?

File: src/lib.rs [hunk 2]
```diff
@@ -421,21 +446,21 @@ impl WasccHost {
...
-            }
-            None => Err(errors::new(errors::ErrorKind::MiscHost(
+        if !self.claims.read().unwrap().contains_key(actor) {
+            return Err(errors::new(errors::ErrorKind::MiscHost(
                 "No such actor".into(),
-            ))),
+            )));
+        }
+        let inv = Invocation::new(
+            "system".to_string(),
```
Comment (ewbankkit) [inline src/lib.rs]: `"system"` is a constant somewhere now I think.

File: src/lib.rs [hunk 3]
```diff
@@ -96,63 +110,48 @@ pub use manifest::{BindingEntry, HostManifest};
...
+    // the key to this field is the subscription subject, and not either a pk or a capid
+    terminators: Arc<RwLock<HashMap<String, Sender<bool>>>>,
     #[cfg(feature = "gantry")]
     gantry_client: Arc<RwLock<Option<gantryclient::Client>>>,
 }
 
 impl WasccHost {
     /// Creates a new waSCC runtime host
-    pub fn new() -> Self {
+    pub fn new() -> WasccHost {
```
Comment (ewbankkit) [inline src/lib.rs]: Why `Self' -> 'WasccHost`?

Comment (autodidaddict) [reply src/lib.rs]: this is leftover from when I was attempting to use generics during a refactor pass. Putting it back to self.

**Comment by autodidaddict** [comment]
Any thoughts on this one @pkedy ?

File: src/bus/mod.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+pub(crate) fn new() -> MessageBus {
+    lattice::DistributedBus::new()
+}
+
+pub(crate) fn actor_subject(actor: &str) -> String {
+    format!("wasmbus.actor.{}", actor)
+}
+
+pub(crate) fn provider_subject(capid: &str, binding: &str) -> String {
+    format!("wasmbus.provider.{}.{}", normalize_capid(capid), binding)
```
Comment (bketelsen) [inline src/bus/mod.rs:41]: perhaps a comment on why normalize_capid is necessary here?

**Review by bketelsen — APPROVED** [review]

File: src/bus/mod.rs (hunk 4 above)
Comment (autodidaddict) [reply src/bus/mod.rs:41]: Added.

**Review by ewbankkit — APPROVED** [review]
LGTM. I want to play with the examples but see nothing to stop this moving forward.

File: src/bus/inproc.rs [hunk 5]
```diff
@@ -0,0 +1,52 @@
+use crate::errors;
```
Comment (ewbankkit) [inline src/bus/inproc.rs:15]: Need to add the

```
Copyright 2015-2020 Capital One Services, LLC
```

header.

File: src/bus/lattice.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
+use crate::{Invocation, InvocationResponse, Result};
```
Comment (ewbankkit) [inline src/bus/lattice.rs:15]: Need to add the

```
Copyright 2015-2020 Capital One Services, LLC
```

header.

File: src/bus/mod.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
+#[cfg(not(feature = "lattice"))]
```
Comment (ewbankkit) [inline src/bus/mod.rs:15]: Need to add the

```
Copyright 2015-2020 Capital One Services, LLC
```

header.

File: src/spawns.rs [hunk 8]
```diff
@@ -0,0 +1,255 @@
+use crate::Result;
```
Comment (ewbankkit) [inline src/spawns.rs:15]: Need to add the

```
Copyright 2015-2020 Capital One Services, LLC
```

header.

**Review by brooksmtownsend — APPROVED** [review]
Looks good, example worked great

**Pull request merged**